lib/db/mysql/index.js

```diff
@@ -2,85 +2,103 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+const path = require('path');
 const mysql = require('mysql');
 const buf = require('buf').hex;
+const MysqlPatcher = require('mysql-patcher');
 
-const AppError = require('../error');
-const config = require('../config');
-const logger = require('../logging')('db.mysql');
-const P = require('../promise');
+const AppError = require('../../error');
+const config = require('../../config');
+const logger = require('../../logging')('db.mysql');
+const P = require('../../promise');
+const patch = require('./patch');
 
-const SCHEMA = require('fs').readFileSync(__dirname + '/schema.sql').toString();
 
 function MysqlStore(options) {
   if (options.charset && options.charset !== 'UTF8_UNICODE_CI') {
-    logger.warn('createDatabase', { charset:  options.charset });
+    logger.warn('createDatabase', { charset: options.charset });
   } else {
     options.charset = 'UTF8_UNICODE_CI';
   }
+  options.typeCast = function(field, next) {
+    if (field.type === 'TINY' && field.length === 1) {
+      return field.string() === '1';
+    }
+    return next();
+  };
   this._pool = mysql.createPool(options);
 }
 
-function createSchema(client, options) {
-  logger.verbose('createSchema', options);
+// Apply patches up to the current patch level.
+// This will also create the DB if it is missing.
+
+function updateDbSchema(patcher) {
+  logger.verbose('updateDbSchema', patcher.options);
 
   var d = P.defer();
-  var database = options.database;
-
-  logger.verbose('createDatabase', database);
-  client.query('CREATE DATABASE IF NOT EXISTS ' + database
-    + ' CHARACTER SET utf8 COLLATE utf8_unicode_ci', function(err) {
-      if (err) {
-        logger.error('createDatabase', err);
-        return d.reject(err);
-      }
+  patcher.patch(function(err) {
+    if (err) {
+      logger.error('updateDbSchema', err);
+      return d.reject(err);
+    }
+    d.resolve();
+  });
 
-      logger.verbose('changeUser');
-      client.changeUser({
-        user: options.user,
-        password: options.password,
-        database: database
-      }, function(err) {
-        if (err) {
-          logger.error('changeUser', err);
-          return d.reject(err);
-        }
-        logger.verbose('creatingSchema');
+  return d.promise;
+}
+
+// Sanity-check that we're working with a compatible patch level.
+
+function checkDbPatchLevel(patcher) {
+  logger.verbose('checkDbPatchLevel', patcher.options);
+
+  var d = P.defer();
+
+  patcher.readDbPatchLevel(function(err) {
+    if (err) {
+      logger.error('checkDbPatchLevel', err);
+      return d.reject(err);
+    }
+    // We are only guaranteed to run correctly if we're at the current
+    // patch level for this version of the code (the normal state of
+    // affairs) or the one immediately above it (during a deployment).
+    if (patcher.currentPatchLevel !== patch.level) {
+      if (patcher.currentPatchLevel !== patch.level + 1) {
```

Comment:
In fxa-auth-db-mysql, we decided to go with it being okay if the DB is at any patch level greater than the one expected. We started with the same as you have here. Just mentioning it here, not expecting you to change it.

I think at one stage it made it easier if there were multiple patches in a release.